src/legacy/core_plugins/interpreter/types/arguments.ts

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { UnwrapPromise } from './common';
+
+/**
+ * This type can convert a type into a known Expression Argument string. For example,
+ * `TypeToArgumentString<Datatable>` will resolve to `'datatable'`.  This allows
+ * Expression Functions to continue to specify their type in a simple string format.
+ */
+export type TypeToArgumentString<T> = KnownTypeToArgumentString<T> | UnmappedArgumentStrings;
+
+/**
+ * This type represents all of the possible combinations of properties of an
+ * Argument in an Expression Function. The presence or absence of certain fields
+ * influence the shape and presence of others within each `arg` in the specification.
+ */
+export type ArgumentType<T> =
+  | SingleArgumentType<T>
+  | MultipleArgumentType<T>
+  | UnresolvedSingleArgumentType<T>
+  | UnresolvedMultipleArgumentType<T>;
+
+/**
+ * Map the type of the generic to a string-based representation of the type.
+ *
+ * If the provided generic is its own type interface, we use the value of
+ * the `type` key as the string literal type for the argument.
+ */
+// prettier-ignore
+type KnownTypeToArgumentString<T> = 
+  T extends string ? 'string' : 
+  T extends boolean ? 'boolean' : 
+  T extends number ? 'number' :
+  T extends null ? 'null' :
+  T extends { type: string } ? T['type'] :
```

Comment:
We can't key off of `type` because sometimes the argument *name* is "type":

https://github.com/elastic/kibana/blob/master/x-pack/plugins/canvas/canvas_plugin_src/functions/server/demodata/index.ts#L16

In addition, this blows up if the type contains a `type` property for a different purpose... consider `Filter`, which has a `type` which refers to the type of filter to be performed:

https://github.com/elastic/kibana/blob/master/x-pack/plugins/canvas/canvas_plugin_src/functions/types/common.ts#L36

We may have to go with `__type`, unfortunately... which lends itself to a whole new set of challenges.